source/web/testing/TestManager.h

```diff
@@ -0,0 +1,274 @@
+        }, std::move(constructor_args));
+      };
+
+      // configure the function that, when called, will call TEST_TYPE::Describe, which should queue
+      // up a Mocha describe function (containing js tests). Next, this function will either (1) queue
+      // up the next test to be run or (2) queue up general TestManager cleanup if there are no more
+      // tests to run.
+      // This function takes advantage of Mocha describe functions to use 'browser' events to chain
+      // together tests (which is necessary because js 'describe' calls add themselves to the browser's
+      // event queue instead of executing as soon as you call them).
+      runner.describe = [runner_id, this]() {
+        this->test_runners[runner_id].test->Describe(); // this will queue up the next test's describe
+
+        const size_t next_test_id = runner_id + 1;
+        auto & test_name = this->test_runners[runner_id].test_name;
+
+        EM_ASM({
+          const test_id = $0;
+          const test_name = UTF8ToString($1);
+
+          // Queue up cleanup for this test
+          describe("Cleanup " + test_name + " (test id " + test_id + ")", function() {
+            it('should clean up test id ' + test_id, function() {
+              emp.CleanupTest($0);
+            });
+          });
+        }, runner_id, test_name.c_str());
+
+        // If there are still more tests to do, queue them
+        // otherwise, queue up a cleanup
+        if (next_test_id < this->test_runners.size()) {
+          auto & next_test_name = this->test_runners[next_test_id].test_name;
+          EM_ASM({
+            const next_test_id = $0;
+            const next_test_name = UTF8ToString($1);
+            // Queue up next test
+            describe("Queue " + next_test_name + " (test id " + next_test_id + ")", function() {
+              it("should queue the next test " + next_test_id, function() {
+                emp.NextTest();
+              });
+            });
+          }, next_test_id, next_test_name.c_str());
+        } else {
+          EM_ASM({
+            describe("Finished running tests.", function() {
+              it("should cleanup test manager", function() {
+                emp.CleanupManager();
+              });
+            });
+          });
+        }
+      };
+
+      // configure the function that, when called, triggers the 'after each test' signal, and cleans
+      // up the dynamically allocated TEST_TYPE object.
+      runner.cleanup = [runner_id, this]() {
+        // Mark test as done.
+        this->test_runners[runner_id].done = true;
+
+        // Did this test trigger any C++ test failures?
+        const size_t post_test_error_cnt = emp::GetUnitTestOutput().errors;
+        auto & test_name = this->test_runners[runner_id].test_name;
+
+        // Did the error count increase after running this test? If so, force failure.
+        if (post_test_error_cnt != this->test_runners[runner_id].before_test_error_count) {
```

Comment:
is the design you're going for to try to end the testing process on the first failure? (I assume, that's what force failure sounds like.) If so, why keep track of before_count? Shouldn't that always be zero if we're failing on the first bad test?